crates/walrus-service/src/node/blob_event_processor.rs

```diff
@@ -0,0 +1,311 @@
+    background_processor_senders: Vec<UnboundedSender<(EventHandle, BlobEvent)>>,
+    _background_processors: Vec<Arc<JoinHandle<()>>>,
+
+    // When there are no background workers, we use a sequential processor to process events using
+    // this processor. This is to keep the same behavior as before BackgroundEventProcessor.
+    sequential_processor: Option<Arc<BackgroundEventProcessor>>,
+}
+
+impl BlobEventProcessor {
+    pub fn new(
+        node: Arc<StorageNodeInner>,
+        blob_sync_handler: Arc<BlobSyncHandler>,
+        num_workers: usize,
+    ) -> Self {
+        let mut senders = Vec::with_capacity(num_workers);
+        let mut workers = Vec::with_capacity(num_workers);
+        for worker_index in 0..num_workers {
+            let (tx, rx) = mpsc::unbounded_channel();
+            senders.push(tx);
+            let mut background_processor =
+                BackgroundEventProcessor::new(node.clone(), blob_sync_handler.clone(), rx);
+            workers.push(Arc::new(tokio::spawn(async move {
+                background_processor.run(worker_index).await;
+            })));
+        }
+
+        let sequential_processor = if num_workers == 0 {
+            // Create a sequential processor to process events sequentially if no background workers
+            // are configured.
+            let (_tx, rx) = mpsc::unbounded_channel();
+            Some(Arc::new(BackgroundEventProcessor::new(
+                node.clone(),
+                blob_sync_handler.clone(),
+                rx,
+            )))
+        } else {
+            None
+        };
+
+        Self {
+            node,
+            background_processor_senders: senders,
+            _background_processors: workers,
+            sequential_processor,
+        }
+    }
+
+    /// Processes a blob event.
+    pub async fn process_event(
+        &self,
+        event_handle: EventHandle,
+        blob_event: BlobEvent,
+    ) -> anyhow::Result<()> {
+        // Update the blob info based on the event.
+        // This processing must be sequential and cannot be parallelized, since there is logical
+        // dependency between events.
+        self.node
+            .storage
+            .update_blob_info(event_handle.index(), &blob_event)?;
+
+        if let BlobEvent::Registered(_) = &blob_event {
+            // Registered event is marked as complete immediately. We need to process registered
+            // events as fast as possible to catch up to the latest event in order to not miss
+            // blob sliver uploads.
+            //
+            // If we want to do this in parallel, we shouldn't mix registered event processing with
+            // certified event processing, as certified events take longer and can block following
+            // registered events.
+            monitored_scope::monitored_scope("ProcessEvent::BlobEvent::Registered");
+            event_handle.mark_as_complete();
```

Comment:
Nit: We could avoid one level of indentation below if we add an `return Ok(());` here.